hyperon-macros/src/lib.rs

```diff
@@ -0,0 +1,524 @@
+        match self {
+            Self::Int(s) => s.clone(),
+            Self::Float(s) => s.clone(),
+            Self::Str(s) => s.clone(),
+            _ => todo!(),
+        }
+    }
+}
+
+struct Tokenizer {
+    state: TokenizerState,
+    input: Box<dyn Iterator<Item=InternalToken>>,
+}
+
+impl Tokenizer {
+    fn new(input: TokenStream) -> Self {
+        let mut prev_end = (0, 0);
+        let input = input.into_iter().flat_map(Self::unroll_group).flat_map(
+            move |it| -> Box<dyn Iterator<Item=InternalToken>> {
+                if prev_end != it.start() {
+                    prev_end = it.end();
+                    Box::new(std::iter::once(InternalToken::Space).chain(std::iter::once(it)))
+                } else {
+                    prev_end = it.end();
+                    Box::new(std::iter::once(it))
+                }
+            });
+        Self {
+            state: TokenizerState::Start,
+            input: Box::new(input),
+        }
+    }
+
+    fn unroll_group(tt: TokenTree) -> Box<dyn Iterator<Item=InternalToken>> {
+        match &tt {
+            TokenTree::Group(g) if g.delimiter() == Delimiter::Parenthesis => {
+                let open = std::iter::once(InternalToken::start_expr(&tt));
+                let close = std::iter::once(InternalToken::end_expr(&tt));
+                Box::new(open.chain(g.stream().into_iter().flat_map(Self::unroll_group)).chain(close))
+            },
+            _ => {
+                Box::new(std::iter::once(InternalToken::TokenTree(tt)))
+            }
+        }
+    }
+
+    fn next(&mut self) -> Token {
+        type TS = TokenizerState;
+        type IT = InternalToken;
+        type T = Token;
+        loop {
+            let (state, it) = match std::mem::replace(&mut self.state, TS::Start) {
+                TS::Token(t) => (TS::Start, Some(t)),
+                state => (state, self.input.next()),
+            };
+            let (token, state) = match (state, it) {
+                (TS::Start, None) => return T::End,
+                (TS::Start, Some(IT::ExprStart(_, _))) => (Some(T::ExprStart), TS::Start),
+                (TS::Start, Some(IT::ExprEnd(_, _))) => (Some(T::ExprEnd), TS::Start),
+                (TS::Start, Some(IT::TokenTree(tt))) => {
+                    match tt {
+                        TokenTree::Literal(l) => {
+                            let s = l.to_string();
+                            let lit = litrs::Literal::parse(s.clone()).expect("Failed to parse literal");
+                            match lit {
+                                litrs::Literal::Integer(_) => (None, TS::Gnd(s, GndType::Int)),
+                                litrs::Literal::Float(_) => (None, TS::Gnd(s, GndType::Float)),
+                                litrs::Literal::String(_) => (None, TS::Gnd(s, GndType::Str)),
+                                litrs::Literal::Bool(_) => (None, TS::Symbol(s)),
+                                _ => (None, TS::Symbol(s)), 
```

Comment:
Since result for both `litrs::Literal::Bool(_)` and `_` is the same maybe it is better to unite them just by removing line `litrs::Literal::Bool(_) => (None, TS::Symbol(s)),`? Or is this intended if, in future, behavior for Bool will change?

Reply:
Fixed, thanks!